hazelcast/src/main/java/com/hazelcast/internal/tpc/ClientAsyncSocketReader.java

```diff
@@ -18,74 +18,106 @@
+        clientMessageReader = new ClientMessageReader(maxMessageLength);
     }
 
     @Override
     public void onRead(ByteBuffer src) {
-        // Currently we just consume the protocol bytes; we don't do anything with it.
         if (!protocolBytesReceived) {
-            consumeProtocolBytes(src);
+            if (!consumeProtocolBytes(src)) {
+                // Not all protocol bytes are available yet
+                return;
+            }
         }
 
         for (; ; ) {
-            if (!clientMessageReader.readFrom(src, true)) {
+            if (!clientMessageReader.readFrom(src, trusted)) {
                 return;
             }
 
             ClientMessage message = clientMessageReader.getClientMessage();
-
             clientMessageReader.reset();
 
             if (connection == null) {
                 loadConnection(message);
-                // now we need to install the socket on the connection
-            } else {
-                message.setConnection(connection);
-                message.setAsyncSocket(socket);
-                clientEngine.accept(message);
             }
+
+            message.setConnection(connection);
+            message.setAsyncSocket(socket);
+            clientEngine.accept(message);
         }
     }
 
-    private void loadConnection(ClientMessage message) {
-        UUID clientUUID = FixedSizeTypesCodec.decodeUUID(message.getStartFrame().content, 0);
-        ClientEndpoint clientEndpoint = findClientEndpoint(clientUUID);
-        if (clientEndpoint == null) {
-            throw new IllegalStateException("Could not find connection for client-uuid:" + clientUUID);
+    private boolean consumeProtocolBytes(ByteBuffer buffer) {
+        if (buffer.remaining() < PROTOCOL_BYTES.length) {
+            // Not enough bytes read
+            return false;
+        }
+
+        if (buffer.get() != PROTOCOL_BYTES[0] || buffer.get() != PROTOCOL_BYTES[1] || buffer.get() != PROTOCOL_BYTES[2]) {
+            throw new IllegalStateException("Received unexpected protocol bytes over socket " + socket);
         }
-        connection = clientEndpoint.getConnection();
-    }
 
-    private void consumeProtocolBytes(ByteBuffer buffer) {
-        // Note(sasha) : AFAIU, it's a trick to reduce garbage production. One object is better than 3.
-        StringBuilder sb = new StringBuilder();
-        sb.append((buffer.get() + buffer.get() + buffer.get()));
         protocolBytesReceived = true;
+        return true;
+    }
+
+    private void loadConnection(ClientMessage message) {
+        if (message.getMessageType() != ExperimentalTpcAuthenticationCodec.REQUEST_MESSAGE_TYPE) {
+            throw new IllegalStateException("Illegal attempt to use " + socket + " before authentication");
+        }
+
+        ExperimentalTpcAuthenticationCodec.RequestParameters request
+                = ExperimentalTpcAuthenticationCodec.decodeRequest(message);
```

Comment:
What if the first message sent is not an authentication message?

Reply:
I did not understand why we have `ExperimentalTpcAuthenticationMessageTask` vs this code. As I know message tasks are run when a member receives a client message. Why have this logic then?

Reply:
1. See the line above this. It is rejected immediately. 

2. Yes, I am aware of this ugliness. The reason is that message tasks are associated with connections, but here, we can't know which connection we should associate to, before parsing the first message. From that message, we learn about the connection. Now, the reason I also do the token check here is that, if I do it later and associate this async socket with that client connection, the next message that comes after this might be executed before I do the token check. I will try to work on this in the 5.4 cycle to adapt message tasks to TPC as well. The reason we have `ExperimentalTpcAuthenticationMessageTask ` is to send response